python/thunder/utils/decorators.py

```diff
@@ -0,0 +1,281 @@
+""" Useful decorators that are used throughout the library """
+
+def _isNamedTuple(obj):
+  """Heuristic check if an object is a namedtuple."""
+  from collections import namedtuple
+  return hasattr(obj, "_fields") \
+     and hasattr(obj, "_asdict") \
+     and callable(obj._asdict)
+
+def serializable(cls):
+    '''The @serializable decorator can decorate any class to make it easy to store
+    that class in a human readable JSON format and then recall it and recover
+    the original object instance. Classes instances that are wrapped in this
+    decorator gain the serialize() method, and the class also gains a
+    deserialize() static method that can automatically "pickle" and "unpickle" a
+    wide variety of objects like so:
+
+      @serializable
```

Comment:
Would be best to reformat a bit so this shows up as an example when building the docs, should be:

```
Examples
--------
>> @serializable
class Visitor():...
```

as in the numpy documentation (e.g. [here](https://github.com/numpy/numpy/blob/master/numpy/linalg/linalg.py)). I think this might be the first proper "example" anywhere in the codebase, so might need to do some trial and error building the docs to get the formatting right, I'm happy to try that myself after this is merged =)